internal/zinc-core/src/main/scala/sbt/internal/inc/InvalidationProfiler.scala

```diff
@@ -0,0 +1,267 @@
+    }
+
+    private def toStringTableIndices(strings: Iterable[String]): Iterable[Long] =
+      strings.map(toStringTableIndex(_))
+
+    private final var compilationStartNanos: Long = 0L
+    private final var compilationDurationNanos: Long = 0L
+    def timeCompilation(startNanos: Long, durationNanos: Long): Unit = {
+      compilationStartNanos = startNanos
+      compilationDurationNanos = durationNanos
+    }
+
+    private def toPathStrings(files: Iterable[File]): Iterable[String] =
+      files.map(_.getAbsolutePath)
+
+    def toApiChanges(changes: APIChanges): Iterable[zprof.ApiChange] = {
+      def toUsedNames(names: Iterable[UsedName]): Iterable[zprof.UsedName] = {
+        import scala.collection.JavaConverters._
+        names.map { name =>
+          val scopes = name.scopes.asScala.map {
+            case UseScope.Default      => zprof.Scope(toStringTableIndex("default"))
+            case UseScope.Implicit     => zprof.Scope(toStringTableIndex("implicit"))
+            case UseScope.PatMatTarget => zprof.Scope(toStringTableIndex("patmat target"))
+          }
+          zprof.UsedName(toStringTableIndex(name.name), scopes.toList)
+        }
+      }
+
+      changes.apiChanges.map {
+        case change: APIChangeDueToMacroDefinition =>
+          zprof.ApiChange(
+            toStringTableIndex(change.modifiedClass),
+            "API change due to macro definition."
+          )
+        case change: TraitPrivateMembersModified =>
+          zprof.ApiChange(
+            toStringTableIndex(change.modifiedClass),
+            s"API change due to existence of private trait members in modified class."
+          )
+        case NamesChange(modifiedClass, modifiedNames) =>
+          val usedNames = toUsedNames(modifiedNames.names).toList
+          zprof.ApiChange(
+            toStringTableIndex(modifiedClass),
+            s"Standard API name change in modified class.",
+            usedNames = usedNames
+          )
+      }
+    }
+
+    private final var initial: Option[zprof.InitialChanges] = None
+    def registerInitial(changes: InitialChanges): Unit = {
+      import scala.collection.JavaConverters._
+      val fileChanges = changes.internalSrc
+      val profChanges = zprof.Changes(
+        added = toStringTableIndices(toPathStrings(fileChanges.getAdded.asScala)).toList,
+        removed = toStringTableIndices(toPathStrings(fileChanges.getRemoved.asScala)).toList,
+        modified = toStringTableIndices(toPathStrings(fileChanges.getChanged.asScala)).toList
+      )
+      initial = Some(
+        zprof.InitialChanges(
+          changes = Some(profChanges),
+          removedProducts = toStringTableIndices(toPathStrings(changes.removedProducts)).toList,
+          binaryDependencies = toStringTableIndices(toPathStrings(changes.binaryDeps)).toList,
+          externalChanges = toApiChanges(changes.external).toList
+        )
+      )
+    }
+
+    private final var currentEvents: List[zprof.InvalidationEvent] = Nil
```

Comment:
Same as with `runs`. This grows unbounded, right?

Reply:
What do you mean by this grows unbounded? In every run, we create a new `RunProfiler` so that new profile wont' see events from the past

Reply:
Ah, thanks :+1: Misunderstood that.